### PR TITLE
python36: 3.6.12 -> 3.6.13; python37: 3.7.9 -> 3.7.10; python27/38/39/310: fix CVE-2021-3177

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

Fixes: CVE-2021-3177

https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-3177
https://bugs.python.org/issue42938

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ajs124** [comment]
What do we want to do about python27? I know it's dead and EOL and we should maybe start an organized effort to try and get rid of it, where it's not needed, but before that, do we want to patch this?

Upstream does not seem to provide patches anymore(?), but various other distros seem to be backporting fixes like this one, e.g. gentoo has a patch for it here: https://gitweb.gentoo.org/fork/cpython.git/commit/?h=gentoo-2.7-vanilla&id=fab838b2ee7cfb9037c24f0f18dfe01aa379b3f7

**Comment by mweinelt** [comment]
Thanks for referencing the patch, I'll try to apply it and we'll see how it goes.

**Comment by SuperSandro2000** [comment]
> What do we want to do about python27? I know it's dead and EOL and we should maybe start an organized effort to try and get rid of it, where it's not needed, but before that, do we want to patch this?

See #101964

> Upstream does not seem to provide patches anymore(?)

No, it does not.

**Comment by fabaff** [comment]
If you want to check your installation:

```bash
[nix-shell:~]$ python
Python 3.8.7 (default, Dec 21 2020, 17:18:55) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> ctypes.c_double.from_param(1e300)
*** buffer overflow detected ***: terminated
Aborted (core dumped
```

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113659` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
7 packages built:
</summary>
<ul>
<li>home-assistant</li>
<li>python27</li>
<li>python310</li>
<li>python36</li>
<li>python37</li>
<li>python38</li>
<li>python39</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
python27:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/usr/include/' doesn't match anything in file 'Lib/plat-atheos/regen'
substituteStream(): WARNING: pattern '/usr/include/' doesn't match anything in file 'Lib/plat-beos5/regen'
substituteStream(): WARNING: pattern '/usr/include/' doesn't match anything in file 'Lib/plat-os2emx/regen'
```

Please check the offending substituteInPlace for typos or changes in source.

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:107:7:

```
    |
107 |       ./use-correct-tcl-tk-on-darwin.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:121:7:

```
    |
121 |       ./2.5.2-ctypes-util-find_library.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:122:7:

```
    |
122 |       ./2.5.2-tkinter-x11.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:123:7:

```
    |
123 |       ./2.6.2-ssl-threads.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:124:7:

```
    |
124 |       ./2.6.5-export-PySignal_SetWakeupFd.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:125:7:

```
    |
125 |       ./2.6.5-FD_SETSIZE.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:126:7:

```
    |
126 |       ./2.6.5-ncurses-abi6.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:127:7:

```
    |
127 |       ./2.7.3-dbm.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:128:7:

```
    |
128 |       ./2.7.3-dylib.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:129:7:

```
    |
129 |       ./2.7.3-getpath-exe-extension.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:130:7:

```
    |
130 |       ./2.7.3-no-libm.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/interpreters/python/cpython/2.7/default.nix:140:7:

```
    |
140 |       ./cross-compile.patch
    |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md
</details>
python310:
<details>


A likely typo in `CPPFLAGS` argument was found, did you mean `CFLAGS` or `CXXFLAGS`?

Near pkgs/development/interpreters/python/cpython/default.nix:240:3:

```
    |
240 |   CPPFLAGS = concatStringsSep " " (map (p: "-I${getDev p}/include") buildInputs);
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
</details>
python36:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern ''/bin/sh'' doesn't match anything in file 'Lib/subprocess.py'
```

Please check the offending substituteInPlace for typos or changes in source.

A likely typo in `CPPFLAGS` argument was found, did you mean `CFLAGS` or `CXXFLAGS`?

Near pkgs/development/interpreters/python/cpython/default.nix:240:3:

```
    |
240 |   CPPFLAGS = concatStringsSep " " (map (p: "-I${getDev p}/include") buildInputs);
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
</details>
python37:
<details>


A likely typo in `CPPFLAGS` argument was found, did you mean `CFLAGS` or `CXXFLAGS`?

Near pkgs/development/interpreters/python/cpython/default.nix:240:3:

```
    |
240 |   CPPFLAGS = concatStringsSep " " (map (p: "-I${getDev p}/include") buildInputs);
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
</details>
python38:
<details>


A likely typo in `CPPFLAGS` argument was found, did you mean `CFLAGS` or `CXXFLAGS`?

Near pkgs/development/interpreters/python/cpython/default.nix:240:3:

```
    |
240 |   CPPFLAGS = concatStringsSep " " (map (p: "-I${getDev p}/include") buildInputs);
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
</details>
python39:
<details>


A likely typo in `CPPFLAGS` argument was found, did you mean `CFLAGS` or `CXXFLAGS`?

Near pkgs/development/interpreters/python/cpython/default.nix:240:3:

```
    |
240 |   CPPFLAGS = concatStringsSep " " (map (p: "-I${getDev p}/include") buildInputs);
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
</details>
</details>

**Review by andir — APPROVED** [review]
Looks good to me. Was able to rebuild all of the python interpreters.

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 1]
```diff
@@ -225,7 +225,7 @@ in with passthru; stdenv.mkDerivation {
   ] ++ optionals (isPy36) [
     # Backport a fix for ctypes.util.find_library.
     ./3.6/find_library.patch
-  ] ++ optionals (isPy38 || isPy39 || isPy310) [
+  ] ++ optionals (isPy27 || isPy38 || isPy39 || isPy310) [
```
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/default.nix:228]: This expression is not used by Python 2.

**Review by FRidh — CHANGES_REQUESTED** [review]
3.8.8 and 3.9.2 have been released now. I'll push those changes and partially cherry-pick the fix for 2.7.

**Comment by mario-canva** [comment]
Hey @FRidh do you have a following up PR for this? Sorry, I couldn't find it and I want to follow it to know once it is merged 😄

**Comment by FRidh** [comment]
Its in staging-next. Now it needs to be cherry-picked to 20.09.

**Comment by erictapen** [comment]
Was backported in 5b37a362d6189acecf51cf27e97689bbf808085a 1b817a0fe7a816565d74e2942e23d6ce2dc8fcab 999d196756eb5d7648f17f98a1ab5546206a8bab.

319940b21c32369a10f2b9fd37113b0daef89890 doesn't need a backport, as CVE-2021-3177 doesn't apply to our python3 versions anymore.

**Pull request closed without merging**